### PR TITLE
[Core] Fix the optimization for IP fetching in `sky launch`

### PR DESCRIPTION
**Pull request by Michaelvll**
The previous optimization for IP fetching does not work as we did not retrieve the internal IP from the provision output, causing the `sky launch` will always fetch the IP addresses causing additional overheads.

This PR fixes the optimization. The following is the profiling, which shows `~9s` faster for launching a new cluster and `~13s` faster for `sky launch` an existing cluster. Note the optimization is for single node only at the moment, but since many users are using single node, it should be helpful.

Profiling (average over 5 runs, on GCP with 2 CPUs, `time sky launch -y -d`):
- Original:
  `sky launch` a new cluster: 2m 41s
  `sky launch` an existing cluster: 1m 26s
- Current PR:
  `sky launch` a new cluster: 2m 32s
  `sky launch` an existing cluster: 1m 13s

TODO:
- [x] Test TPU pod

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py`
- [x] All smoke tests: `pytest tests/test_smoke.py --aws`
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -2374,29 +2374,31 @@ def _internal_ips(self) -> Optional[List[str]]:
...
-            max_attempts: int = _FETCH_IP_MAX_ATTEMPTS) -> Optional[List[str]]:
+    def internal_ips(self,
+                     max_attempts: int = _FETCH_IP_MAX_ATTEMPTS) -> List[str]:
         internal_ips = self._internal_ips
         if internal_ips is not None:
             return internal_ips
         self.update_cluster_ips(max_attempts=max_attempts)
-        return self._internal_ips
+        internal_ips = self._internal_ips
+        assert internal_ips is not None, 'update_cluster_ips failed.'
```
Comment (cblmemo) [inline sky/backends/cloud_vm_ray_backend.py:2384]: Shall we raise an error here?

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -2374,29 +2374,31 @@ def _internal_ips(self) -> Optional[List[str]]:
...
+    def external_ips(self,
+                     max_attempts: int = _FETCH_IP_MAX_ATTEMPTS) -> List[str]:
         external_ips = self._external_ips
         if external_ips is not None:
-            return self._external_ips
+            return external_ips
         self.update_cluster_ips(max_attempts=max_attempts)
-        return self._external_ips
+        external_ips = self._external_ips
+        assert external_ips is not None, 'update_cluster_ips failed.'
```
Comment (cblmemo) [inline sky/backends/cloud_vm_ray_backend.py:2400]: ditto

File: sky/backends/backend_utils.py [hunk 3]
```diff
@@ -1995,22 +1995,11 @@ def run_ray_status_to_check_ray_cluster_healthy() -> bool:
...
-            # resources handle is written back to local database.
-            # This is helpful when user interrupt after the provision is done
-            # and before the skylet is restarted. After #2304 is merged, this
-            # helps keep the cluster status to INIT after `sky status -r`, so
-            # user will be notified that any auto stop/down might not be
-            # triggered.
-            if external_ips is None or len(external_ips) == 0:
-                raise exceptions.FetchIPError(
-                    reason=exceptions.FetchIPError.Reason.HEAD)
+            # TODO(zhwu): check the correctness of stopped TPU VM
```
Comment (cblmemo) [inline sky/backends/backend_utils.py:2001]: Shall we add a TODO for the case in #2304 ?

**Review by cblmemo — APPROVED** [review]
The code looks good to me! Left several nits:

**Comment by Michaelvll** [comment]
I fixed some problems with the TPU pod. It now passes all smoke tests for GCP and AWS. PTAL @cblmemo.

File: sky/backends/cloud_vm_ray_backend.py [hunk 4]
```diff
@@ -2238,38 +2294,85 @@ def _update_cluster_region(self):
...
 
         self.launched_resources = self.launched_resources.copy(region=region)
 
-    def _update_stable_ssh_ports(self, max_attempts: int = 1) -> None:
+    def update_ssh_ports(self, max_attempts: int = 1) -> None:
         # TODO(romilb): Replace this with a call to the cloud class to get ports
-        if isinstance(self.launched_resources.cloud, clouds.Kubernetes):
-            head_port = backend_utils.get_head_ssh_port(
-                self, use_cache=False, max_attempts=max_attempts)
-            # TODO(romilb): Multinode doesn't work with Kubernetes yet.
```
Comment (cblmemo) [inline sky/backends/cloud_vm_ray_backend.py:2251]: Shall we keep this TODO?

**Review by cblmemo — COMMENTED** [review]
The code looks great to me! Left one nit and two questions that I'm not entirely sure I understand:
* Why do we need manually pass `port` argument when initializing `SSHCommandRunner`?
* I noticed that our IP is ephemeral. Is there a possibility that `prev_handle` might contain stale IP if we `sky stop` some `UP` cluster?

**Comment by Michaelvll** [comment]
Thanks for the review @cblmemo!
> Why do we need manually pass port argument when initializing SSHCommandRunner?

The main reason is for the type checking. It seems if we do not manually pass that argument, `mypy` will complain about the `**ssh_crednetials` is `**Dict[str, str]` but the remaining arguments (i.e., port) in that function has `int` type.

> I noticed that our IP is ephemeral. Is there a possibility that prev_handle might contain stale IP if we sky stop some UP cluster?

It is ok to pass the staled IP, because we will update the IPs later in the code once the VM is provisioned. https://github.com/skypilot-org/skypilot/blob/95ac2313f91decfce1bffc1b2eeb92a993e8f825/sky/backends/cloud_vm_ray_backend.py#L1596

And in the update, we will check if the IP from the `ray up` matches the one in the cache. If they do not match, we will use the new one and update the internal IPs.

**Review by cblmemo — APPROVED** [review]
Looks great to me!

**Pull request merged**